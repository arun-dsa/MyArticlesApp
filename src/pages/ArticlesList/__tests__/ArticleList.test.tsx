import { screen, render, waitFor } from "@testing-library/react";
import { setupServer } from "msw/node";
import { afterAll, afterEach, beforeAll, describe, expect, test } from "vitest";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

import { routeWrapper, apiHandlers } from "@utils/test_utils";
import { ARTICLES } from "@tests/mockData";
import ArticleList from "../index";

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});

describe("Article List test", () => {
  const server = setupServer(apiHandlers()?.getAllArticle());

  beforeAll(() => server.listen());

  afterEach(() => {
    server.resetHandlers();
    queryClient.clear();
  });

  afterAll(() => server.close());

  test("Article List screen render", async () => {
    const { container } = render(
      <QueryClientProvider client={queryClient}>
        {routeWrapper([{ path: "/", element: <ArticleList /> }])}
      </QueryClientProvider>
    );

    expect(screen.getAllByTestId("article-card-skeleton")).toHaveLength(10);

    ARTICLES?.forEach((article) =>
      expect(
        screen.queryByTestId(`article-card-${article.id}`)
      ).not.toBeInTheDocument()
    );

    await screen.findByTestId(`article-card-${ARTICLES[0].id}`);

    ARTICLES?.map((article) =>
      expect(
        screen.queryByTestId(`article-card-${article.id}`)
      ).toBeInTheDocument()
    );

    expect(container).toMatchSnapshot();
  });
});

describe("Article List test with Error", () => {
  const server = setupServer(apiHandlers()?.getAllArticleWithError());

  beforeAll(() => server.listen());

  afterEach(() => {
    server.resetHandlers();
    queryClient.clear();
  });

  afterAll(() => server.close());

  test("Error screen render", async () => {
    const { container } = render(
      <QueryClientProvider client={queryClient}>
        {routeWrapper([{ path: "/", element: <ArticleList /> }])}
      </QueryClientProvider>
    );

    expect(screen.getAllByTestId("article-card-skeleton")).toHaveLength(10);

    ARTICLES?.forEach((article) =>
      expect(
        screen.queryByTestId(`article-card-${article.id}`)
      ).not.toBeInTheDocument()
    );

    await screen.findByTestId("error-page");

    expect(screen.getByText("Something went wrong")).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });
});
