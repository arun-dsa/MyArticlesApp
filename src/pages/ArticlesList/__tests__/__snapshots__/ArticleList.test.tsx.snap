// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Article List test > Article List screen render 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 css-11l5t4l-MuiGrid-root"
        data-testid="article-card-1"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1rp1zv7-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-97bign-MuiCardContent-root"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1u4kd2-MuiTypography-root"
              data-testid="article-title-1"
            >
              Exploring the Wonders of the Great Barrier Reef
            </h5>
            <p
              class="MuiTypography-root MuiTypography-body2 css-1i87a3m-MuiTypography-root"
              data-testid="article-summary-1"
            >
              Discover the incredible biodiversity and natural beauty of the Great Barrier Reef.
            </p>
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
          />
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1mgzbsd-MuiCardActions-root"
          >
            <a
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary css-nw1vud-MuiButtonBase-root-MuiButton-root"
              data-testid="share-article-1"
              href="mailto:change-it-to-email-to-share@gmail.com?subject=Read this article - Exploring the Wonders of the Great Barrier Reef&body=Discover the incredible biodiversity and natural beauty of the Great Barrier Reef.. Read the full article here http://localhost:3000/articles/1"
              tabindex="0"
            >
              <span
                class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ShareIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81 1.66 0 3-1.34 3-3s-1.34-3-3-3-3 1.34-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9c-1.66 0-3 1.34-3 3s1.34 3 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.16c-.05.21-.08.43-.08.65 0 1.61 1.31 2.92 2.92 2.92 1.61 0 2.92-1.31 2.92-2.92s-1.31-2.92-2.92-2.92"
                  />
                </svg>
              </span>
              Share
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </a>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorPrimary css-8fujfr-MuiButtonBase-root-MuiButton-root"
              data-testid="article-view-full-1"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ListAltIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5zm1.1-2H3.9c-.5 0-.9.4-.9.9v16.2c0 .4.4.9.9.9h16.2c.4 0 .9-.5.9-.9V3.9c0-.5-.5-.9-.9-.9M11 7h6v2h-6zm0 4h6v2h-6zm0 4h6v2h-6zM7 7h2v2H7zm0 4h2v2H7zm0 4h2v2H7z"
                  />
                </svg>
              </span>
              View Full Article
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 css-11l5t4l-MuiGrid-root"
        data-testid="article-card-2"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1rp1zv7-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-97bign-MuiCardContent-root"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1u4kd2-MuiTypography-root"
              data-testid="article-title-2"
            >
              The Rise of Electric Vehicles: Shaping the Future of Transportation
            </h5>
            <p
              class="MuiTypography-root MuiTypography-body2 css-1i87a3m-MuiTypography-root"
              data-testid="article-summary-2"
            >
              Learn about the advancements in electric vehicle technology and their impact on reducing carbon emissions.
            </p>
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
          />
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1mgzbsd-MuiCardActions-root"
          >
            <a
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary css-nw1vud-MuiButtonBase-root-MuiButton-root"
              data-testid="share-article-2"
              href="mailto:change-it-to-email-to-share@gmail.com?subject=Read this article - The Rise of Electric Vehicles: Shaping the Future of Transportation&body=Learn about the advancements in electric vehicle technology and their impact on reducing carbon emissions.. Read the full article here http://localhost:3000/articles/2"
              tabindex="0"
            >
              <span
                class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ShareIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81 1.66 0 3-1.34 3-3s-1.34-3-3-3-3 1.34-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9c-1.66 0-3 1.34-3 3s1.34 3 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.16c-.05.21-.08.43-.08.65 0 1.61 1.31 2.92 2.92 2.92 1.61 0 2.92-1.31 2.92-2.92s-1.31-2.92-2.92-2.92"
                  />
                </svg>
              </span>
              Share
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </a>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorPrimary css-8fujfr-MuiButtonBase-root-MuiButton-root"
              data-testid="article-view-full-2"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ListAltIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5zm1.1-2H3.9c-.5 0-.9.4-.9.9v16.2c0 .4.4.9.9.9h16.2c.4 0 .9-.5.9-.9V3.9c0-.5-.5-.9-.9-.9M11 7h6v2h-6zm0 4h6v2h-6zm0 4h6v2h-6zM7 7h2v2H7zm0 4h2v2H7zm0 4h2v2H7z"
                  />
                </svg>
              </span>
              View Full Article
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Article List test with Error > Error screen render 1`] = `
<div>
  <div
    class="MuiBox-root css-3z75l1"
    data-testid="error-page"
  >
    <h4
      class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1vw6mcs-MuiTypography-root"
    >
      Something went wrong
    </h4>
    <p
      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    >
      An error occurred while fetching articles. Please try agian later.Response status: 500
    </p>
  </div>
</div>
`;
